sklearn/metrics/_ranking.py

```diff
@@ -696,8 +697,14 @@ def _binary_clf_curve(y_true, y_score, pos_label=None, sample_weight=None):
     assert_all_finite(y_true)
     assert_all_finite(y_score)
 
+    # Filter out zero-weighted samples, as they should not impact the result
     if sample_weight is not None:
         sample_weight = column_or_1d(sample_weight)
+        sample_weight = _check_sample_weight(sample_weight, y_true)
+        nonzero_weight_mask = sample_weight != 0
+        y_true = y_true[nonzero_weight_mask]
+        y_score = y_score[nonzero_weight_mask]
+        sample_weight = sample_weight[nonzero_weight_mask]
```

Comment:
Would it not be better to instead of filtering out zero sample weights, doing the computation of scores in a way that zero sample weights are not taken into account? This proposed change results in a difference between `0.` and epsilon, which I don't think is desired.

Reply:
For whatever reasons you have some zeros in `sample_weight`, you better always get the same length of arrays. Imagine `y_score = clf.predict_proba(X)` with `y_score.shape[0] == X.shape[0] == sample_weight.shape[0]`. Then, filtering out zero weight samples here in `_binary_clf_curve` is quite convenient.

> This proposed change results in a difference between 0. and epsilon, which I don't think is desired.

Could you elaborate?

Reply:
Sorry @adrinjalali I do not understand your point either...

Reply:
let say we want a weighted average of a bunch of numbers, one way is first to filter out data with zero weight, the other is to have sth like: `sum(w[i]*x[i]) / sum(w[i])` and ignore the fact that some of those weights are zero. The calculation itself takes care of zero sample weights.

Now in this case, there's an issue with having samples with zero weight in the data. My question is if we just filter out the zero weights, then what happens to the sample with weight equal to 1e-32 for instance? That's practically zero, and should be treated [almost] the same as the ones with zero sample weight.

Reply:
AFAIK, `_binary_clf_curve` is only used here in _ranking.py. It calculates counts `cumsum(y * weight)` which are the ingredients for precision, recall and the like. You never have `... / sum(weight)`. This would cancel out in ratios of counts anyway. Therefore, I would have expected that filtering out zero weight samples is equivalent.
The real problem mentioned in #16065, however, seems to be the different values of thresholds, i.e. thresholds with only zero weights should be dropped. This PR solves this problem (by removing those sample).

Suggestion: Should we just add a test with `weight1=[1, 1, 1e-30]` and `weight2=[1, 1, 0]` for several relevant scores/erros/losses?

Reply:
Can we reach a consensus? I would like to get this one closed?

Reply:
I played around a little bit, I'm happy with this solution as is :)